packages/autoloader/src/AutoloadGenerator.php

```diff
@@ -313,44 +311,24 @@ private function getAutoloadFilesPackagesFile( $filesMap ) {
 	/**
 	 * Generate the PHP that will be used in the autoload_packages.php files.
 	 *
-	 * @param string $suffix  Unique suffix added to the jetpack_enqueue_packages function.
+	 * @param String $suffix  Unique suffix added to the jetpack_enqueue_packages function.
 	 *
 	 * @return string
 	 */
 	private function getAutoloadPackageFile( $suffix ) {
-		$sourceLoader   = fopen( __DIR__ . '/autoload.php', 'r' );
-		$file_contents  = stream_get_contents( $sourceLoader );
-		$file_contents .= <<<INCLUDE_FILES
-/**
- * Prepare all the classes for autoloading.
- */
-function enqueue_packages_$suffix() {
-	\$class_map = require_once dirname( __FILE__ ) . '/composer/autoload_classmap_package.php';
-	foreach ( \$class_map as \$class_name => \$class_info ) {
-		enqueue_package_class( \$class_name, \$class_info['version'], \$class_info['path'] );
-	}
+		$sourceLoader  = fopen( __DIR__ . '/autoload.php', 'r' );
+		$file_contents = stream_get_contents( $sourceLoader );
+		$file_contents = str_replace(
+			'Automattic\\Jetpack\\Autoloader;',
+			'Automattic\\Jetpack\\Autoloader\\jp' . $suffix . ';',
+			$file_contents
```

Comment:
I'm a little concerned about this `str_replace()`. I think that we'll probably be rearranging the spl autoloader chain in a function that will be in the `autoload.php` file. The old autoloader function that we'll need to check for and move around is `Automattic\Jetpack\Autoloader\autoloader`. So that would be affected by the `str_replace()`.

Reply:
So my understanding on where to go from here based on our discussions was this:
* we're loading every `autoload_packages.php`, generated from new autoloaders, at the time the first new autoloader loads ( :clap: most uses of the word 'load' in a sentence :clap: )
* every time an autoloader file loads, it adds its namespace to a global array, along with its package version.
* after all autoloader files are globbed and loaded, we have a namespaces map, classes map, and files map. From here we can determine which version is the newest and add its `autoload` method using `spl_autoload_register`, making sure it's the first one.

Does this sound logical?

Reply:
I think it’s important that only the latest autoloader be involved in generating the in-memory classmap and filemap. The way that versions are selected for these maps could change as we update the autoloader. For example, the old autoloader always favors dev versions, while the new autoloader will be more restrictive about when to use dev versions. Over time, we might make additional changes to how versions are selected, so we'll want to make sure that the newest autoloader does this work.

So, I think we need to find the latest autoloader version first, and then have it do all the work.

Here’s how I imagined this would work:

1) The first autoloader identifies the autoloader with the latest version. It does this by traversing through the active plugin folders, looking for `jetpack_autoload_classmap.php` files. It gets the autoloader versions from the  'Automattic\\Jetpack\\Autoloader\\AutoloadGenerator' entries in those files. It stores the latest version value in the global `$latest_autoloader_version`, and then it requires the `autoload_packages.php` file for the latest autoloader.

2) The latest autoloader compares the `$latest_autoloader_version` global with its own version, and determines that it is the latest version. Then it does all the work:
  - Generates the classmap and file map using the classmap and file map files from all of the active plugins.
  - Registers it’s autoloader.
  - Adjusts the spl autoloader chain if necessary to make sure that the new autoloader function runs before the old one.

About globbing the files - I was concerned that the site could have inactive plugins, and we wouldn’t want to look at the files for inactive plugins. That's why I suggested to traverse through the active plugin folders to find the files. What do you think?

Reply:
We can disregard my concern about the `str_replace()` call. The autoloader name isn't changed by that call.